### PR TITLE
feat: Use cloud-hypervisor API to start VM

### PR DESCRIPTION
**Pull request by scuzhanglei**

### CODE REVIEW

**Comment by scuzhanglei** [comment]
Because this s6-overlay [issue](https://github.com/just-containers/s6-overlay/issues/477), if an error occurred in virt-prerunner, the pod will still be `running`, instead of being `failed`. this cause VM hang on `scheduled`.

File: cmd/virt-prerunner/main.go [hunk 1]
```diff
@@ -48,88 +48,36 @@ func main() {
...
+			Mode: "Pty",
+		},
+		Serial: &cloudhypervisor.ConsoleConfig{
+			Mode: "Tty",
+		},
 		Payload: &cloudhypervisor.PayloadConfig{
 			Kernel: "/var/lib/cloud-hypervisor/hypervisor-fw",
 		},
 		Cpus: &cloudhypervisor.CpusConfig{
+			MaxVcpus:  int(vm.Spec.Instance.CPU.Sockets * vm.Spec.Instance.CPU.CoresPerSocket),
```
Comment (fengye87) [inline cmd/virt-prerunner/main.go:80]: This is required?

File: pkg/controller/vm_controller.go [hunk 2]
```diff
@@ -279,7 +279,12 @@ func (r *VMReconciler) buildVMPod(ctx context.Context, vm *virtv1alpha1.VirtualM
 				SecurityContext: &corev1.SecurityContext{
 					Privileged: func() *bool { v := true; return &v }(),
 				},
-				Args: []string{"--vm-data", base64.StdEncoding.EncodeToString(vmJSON)},
+				Env: []corev1.EnvVar{
+					{
```
Comment (fengye87) [inline pkg/controller/vm_controller.go:283]: Merge this with the line above

File: pkg/daemon/vm_controller.go [hunk 3]
```diff
@@ -164,6 +164,19 @@ func (r *VMReconciler) reconcile(ctx context.Context, vm *virtv1alpha1.VirtualMa
 			switch vm.Status.Migration.Phase {
 			case virtv1alpha1.VirtualMachineMigrationScheduled:
 				if vm.Status.Migration.TargetNodeName == r.NodeName {
+					if err := wait.PollImmediate(time.Second, 3*time.Second, func() (done bool, err error) {
```
Comment (fengye87) [inline pkg/daemon/vm_controller.go:167]: Is this change related with this issue?

**Review by fengye87 — CHANGES_REQUESTED** [review]

File: cmd/virt-prerunner/main.go (hunk 1 above)
Comment (scuzhanglei) [reply cmd/virt-prerunner/main.go:80]: yes, this field is required, as We don't support hotplug CPU now, so set it to this value.

Comment (fengye87) [reply cmd/virt-prerunner/main.go:80]: But it's not required in cmd args?

File: pkg/daemon/vm_controller.go (hunk 3 above)
Comment (scuzhanglei) [reply pkg/daemon/vm_controller.go:167]: yes, now in virt-prerunner, cloud-hypervisor is running by s6 as a service, when the pod become `running`, cloud-hypervisor maybe maybe is not running and the socket file is not created. so we have to wait it.

File: cmd/virt-prerunner/main.go (hunk 1 above)
Comment (scuzhanglei) [reply cmd/virt-prerunner/main.go:80]: according the [openapi](https://raw.githubusercontent.com/cloud-hypervisor/cloud-hypervisor/v26.0/vmm/src/api/openapi/cloud-hypervisor.yaml) `MaxVcpus` is required.

File: build/virt-prerunner/cloud-hypervisor-type [hunk 4]
```diff
@@ -0,0 +1 @@
+longrun
```
Comment (carezkh) [inline build/virt-prerunner/cloud-hypervisor-type:1]: What if CH crashed? IIUC, type _longruns_ get supervised by s6 and will be restarted if crashed, pod `prerunner` will never get failed.

Currently, errors will be reported in `daemon` controller when VM info not found for a `Running` VM, but i noticed there is a `TODO` that means VM info not found will be ignored in the future.
https://github.com/smartxworks/virtink/blob/ff9679140b2b1b6d298d8eec581b27258b222111/pkg/daemon/vm_controller.go#L97-L101

Comment (scuzhanglei) [reply build/virt-prerunner/cloud-hypervisor-type:1]: as s6 adviced, I add a [finish script](https://github.com/just-containers/s6-overlay#writing-an-optional-finish-script) for cloud-hypervisor service, when it stoped, the finish script will be executed, and in finish script call `/run/s6/basedir/bin/halt` to stop the container.

**Comment by scuzhanglei** [comment]
> Because this s6-overlay [issue](https://github.com/just-containers/s6-overlay/issues/477), if an error occurred in virt-prerunner, the pod will still be `running`, instead of being `failed`. this cause VM hang on `scheduled`.

this issue has been fixed by adding `S6_BEHAVIOUR_IF_STAGE2_FAILS=2` env to the container, it will stop the container if any error occurred in virt-prerunner.

**Comment by scuzhanglei** [comment]
merged in https://github.com/smartxworks/virtink/pull/73
